### DISCUSSION TITLE
SQLite and WAL usage

### DISCUSSION THREAD

**Discussion by xmichaelx** [directus/directus#14153] [Archive]
Is Directus using WAL with SQLite backend? If not is it possible to enable it with executing `pragma journal_mode=wal;` somewhere? Can it be configured?

I see that in https://github.com/directus/directus/blob/895100898cd6f0f6121e964e793137059e82fc1b/api/src/database/index.ts
there is line which enables foreign_keys `await run('PRAGMA foreign_keys = ON');` would it be possible to add WAL? or at least do so depending on additional variable?

**Comment by rijkvanzanten** [comment]
Directus currently does not enable WAL, and there's currently no configuration option to enable it. 
@azrikahar could your spatialite injection hack work for this use case?

**Reply by azrikahar** [reply]
Not too sure myself tbh, but it _seems_ to work when I try it out locally 😮 

1. Create a hook extension (we are mainly accessing the knex instance via `database`, and it's not technically a hook, hence the "hack" pretense):

   ```js
   module.exports = async (_, { database, logger }) => {
	 // Skip the rest of logic if it's not sqlite3
	 if (database.client.config.client !== 'sqlite3') return;

	 const pool = database.client.pool;
	 const acquire = pool.acquire();

	 try {
	 	const conn = await acquire.promise;
		await conn.run('PRAGMA journal_mode = WAL;');
		logger.info('WAL mode turned on successfully.');
		pool.release(conn);
	 } catch (err) {
		logger.error('Failed to turn on WAL mode.');
 	 }
   };
   ```

2. It should work now. But optionally I did the following custom endpoint just as a way to do some form of sanity check:

   ```js
   module.exports = (router, { database }) => {
	 router.get('/', (req, res) => {
		 database.raw('PRAGMA journal_mode;').then((r) => {
			 res.send(r);
		 });
	 });
   };
   ```
 
   |Before|After|
   |---|---|
   |![chrome_gjRFhjDli0](https://user-images.githubusercontent.com/42867097/176333247-c1122b4e-e3bf-4dc5-a413-1e3e1e91fbd0.png)|![chrome_UKQAeBVErq](https://user-images.githubusercontent.com/42867097/176333295-f708daeb-d326-4cb8-8923-328b375df504.png)|

**Reply by xmichaelx** [reply]
Awesome, thank you. This is very useful especially with very local deployment of Directus where it serves as view on local SQLite database which is being used as storage by other processes. Another good thing to set for local deployments on top of SQLite is ability for the process pool to go down to 0 after short period of inactivity since pooling connections for local SQLite database makes little sense. I use following additional environment options:
```
DB_POOL_MIN=0
DB_POOL_idleTimeoutMillis=2000
```

This helped when I had process which was writing into the database opened in directus using WAL, cause without it WAL file was just getting bigger because Directus was holding connection on the database and there was no moment when WAL was able to be merged into the db.

Perhaps they along with WAL should be the default options in Directus operating on top of SQLite? Of course I understand if you don't want to change anything in storage that is meant to be mainly for demonstration purposes :)

**Reply by rijkvanzanten** [reply]
I'm a little torn on whether we should change defaults based on driver used, or document the global defaults better, and add a "TIP" note on _recommended_ configuration for SQLite 🤔 Changing default behavior based on other configuration flags feels harder to explain to me

**Reply by rijkvanzanten** [reply]
In any case, lets add a dedicated `init` hook for the `afterCreate` on the pool event @azrikahar, so we can start recommending your solution as the official way to tap in custom extensions or other settings on the database connection, rather than having to rely on additional individual options for every permutation 🙂 https://github.com/directus/directus/blob/e386caabc1659a8c42c9c09f60390dd61d403caa/api/src/database/index.ts#L91

**Reply by rijkvanzanten** [reply]
I've created an actionable ticket for that improvement here: https://github.com/directus/directus/issues/14310 🙂

**Reply by xmichaelx** [reply]
That + example on how to use that in documentation solves the issue for me - and it looks like something other drivers could also benefit from so that's a nice synergy.
